File: DailyChallenges/Scala/LeetCode_974.scala
object Solution {
    def subarraysDivByK(nums: Array[Int], k: Int): Int = {
        val cnt = Array.ofDim[Int](k); cnt(0) = 1
        val (_, ans) = nums.foldLeft((0,0))((p, n) => {
            val prefix = (p._1 + n % k + k) % k
            cnt(prefix) = cnt(prefix) + 1
            (prefix, p._2 + cnt(prefix)-1)
        })
        ans
    }
}

File: DailyChallenges/CPP/LeetCode_974.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        vector<int> cnt(k); cnt[0] = 1;
        int prefix = 0, ans = 0;
        for (int n: nums){
            prefix = (prefix + n%k +k) % k;
            ans += cnt[prefix]++;
        }
        return ans;
    }
};
